zmq/eventloop/future.py

```diff
@@ -0,0 +1,309 @@
+"""Future-returning APIs for coroutines."""
+
+# Copyright (c) PyZMQ Developers.
+# Distributed under the terms of the Modified BSD License.
+
+from collections import namedtuple
+
+try:
+    from concurrent.futures import Futurexx
+except:
+    from tornado.concurrent import Future as TFuture
+    
+    class CancelledError(Exception):
+        pass
+    
+    class Future(TFuture):
+        """Subclass Tornado Future, reinstating cancellation."""
+        def cancel(self):
+            if self.done():
+                return False
+            self.set_exception(CancelledError())
+            return True
+        
+        def cancelled(self):
+            return self.done() and isinstance(self.exception(), CancelledError)
+
+import zmq as _zmq
+from zmq.eventloop.ioloop import IOLoop
+
+
+_FutureEvent = namedtuple('_FutureEvent', ('future', 'kind', 'args', 'msg'))
+
+
+class Poller(_zmq.Poller):
+    """Poller that returns a Future on poll, instead of blocking."""
+    def poll(self, timeout=-1):
+        """Return a Future for a poll event"""
+        future = Future()
+        if timeout == 0:
+            try:
+                result = super(Poller, self).poll(0)
+            except Exception as e:
+                future.set_exception(e)
+            else:
+                future.set_result(result)
+            return future
+        
+        loop = IOLoop.current()
+        
+        # register Future to be called as soon as any event is available on any socket
+        # only support polling on zmq sockets, for now
+        watcher = Future()
+        for socket, mask in self.sockets:
+            if mask & _zmq.POLLIN:
+                socket._add_recv_event('poll', future=watcher)
+            if mask & _zmq.POLLOUT:
+                socket._add_send_event('poll', future=watcher)
+        
+        def on_poll_ready(f):
+            if future.done():
+                return
+            if watcher.exception():
+                future.set_exception(watcher.exception())
+            else:
+                try:
+                    result = super(Poller, self).poll(0)
+                except Exception as e:
+                    future.set_exception(e)
+                else:
+                    future.set_result(result)
+        watcher.add_done_callback(on_poll_ready)
+        
+        if timeout > 0:
+            # schedule cancel to fire on poll timeout, if any
+            def trigger_timeout():
+                if not watcher.done():
+                    watcher.set_result(None)
+            
+            timeout_handle = loop.call_later(
+                1e-3 * timeout,
+                trigger_timeout
+            )
+            def cancel_timeout(f):
+                loop.remove_timeout(timeout_handle)
+            future.add_done_callback(cancel_timeout)
+        
+        def cancel_watcher(f):
+            if not watcher.done():
+                watcher.cancel()
+        future.add_done_callback(cancel_timeout)
```

Comment:
is `cancel_timeout` a typo here? Did you actually mean `future.add_done_callback(cancel_watcher)`?